### PR TITLE
adds allowed_client_ids field to identity token named keys

### PR DESCRIPTION
**Pull request by Lexman42**

### CODE REVIEW

**Comment by jefferai** [comment]
I think this actually needs to be allowed client IDs. We and @kalafut should discuss.

File: vault/identity_store_oidc.go [hunk 1]
```diff
@@ -140,6 +135,11 @@ func oidcPaths(i *IdentityStore) []*framework.Path {
 					Description: "Signing algorithm to use. This will default to RS256.",
 					Default:     "RS256",
 				},
+
+				"allowed_clientIDs": &framework.FieldSchema{
```
Comment (kalafut) [inline vault/identity_store_oidc.go]:
```suggestion
				"allowed_client_ids": &framework.FieldSchema{
```

File: vault/identity_store_oidc.go [hunk 2]
```diff
@@ -751,15 +766,25 @@ func (i *IdentityStore) pathOIDCCreateUpdateRole(ctx context.Context, req *logic
 		return logical.ErrorResponse("key must be provided"), nil
```
Comment (kalafut) [inline vault/identity_store_oidc.go]: Weren't we going to remove this check?

File: vault/identity_store_oidc.go [hunk 3]
```diff
@@ -44,17 +39,17 @@ type expireableKey struct {
...
 
 type namedKey struct {
-	Name            string           `json:"name"`
-	Algorithm       string           `json:"signing_algorithm"`
-	VerificationTTL time.Duration    `json:"verification_ttl"`
-	RotationPeriod  time.Duration    `json:"rotation_period"`
-	KeyRing         []*expireableKey `json:"key_ring"`
-	SigningKey      *jose.JSONWebKey `json:"signing_key"`
-	NextRotation    time.Time        `json:"next_rotation"`
+	Name             string           `json:"name"`
```
Comment (kalafut) [inline vault/identity_store_oidc.go]: Can this field be removed now?

File: vault/identity_store_oidc.go [hunk 4]
```diff
@@ -751,15 +766,25 @@ func (i *IdentityStore) pathOIDCCreateUpdateRole(ctx context.Context, req *logic
...
 	}
 
+	// Validate that the role is allowed to use this key, if it is not, warn the user
+	var namedKey namedKey
+	warning := ""
+	if err := entry.DecodeJSON(&namedKey); err != nil {
+		return nil, err
+	}
+	if !strutil.StrListContains(namedKey.AllowedClientIDs, "*") && !strutil.StrListContainsGlob(namedKey.AllowedClientIDs, name) {
+		warning = fmt.Sprintf("The key %q does not list the client id of the role %q as an allowed_clientID and needs to be updated before ID tokens can be generated against this role.", role.Key, name)
```
Comment (kalafut) [inline vault/identity_store_oidc.go]:
```suggestion
		warning = fmt.Sprintf("The key %q does not list the client ID of the role %q as an allowed client ID.", role.Key, name)
```

File: vault/identity_store_oidc.go [hunk 5]
```diff
@@ -358,16 +358,19 @@ func (i *IdentityStore) pathOIDCCreateUpdateKey(ctx context.Context, req *logica
 	name := d.Get("name").(string)
 
 	var key namedKey
-	entry, err := req.Storage.Get(ctx, namedKeyConfigPath+name)
-	if err != nil {
-		return nil, err
+	if req.Operation == logical.CreateOperation {
+		key.name = name
```
Comment (kalafut) [inline vault/identity_store_oidc.go]: Is this needed?

File: vault/identity_store_oidc_test.go [hunk 6]
```diff
@@ -671,7 +723,7 @@ func TestOIDC_pathOIDCKeyExistenceCheck(t *testing.T) {
 
 	// Populte storage with a namedKey
 	namedKey := &namedKey{
-		Name: keyName,
+		name: keyName,
```
Comment (kalafut) [inline vault/identity_store_oidc_test.go]: Not needed

File: vault/identity_store_oidc_test.go [hunk 7]
```diff
@@ -732,9 +784,7 @@ func TestOIDC_pathOIDCRoleExistenceCheck(t *testing.T) {
 	}
 
 	// Populte storage with a role
```
Comment (kalafut) [inline vault/identity_store_oidc_test.go]:
```suggestion
	// Populate storage with a role
```

**Pull request merged**